Add color support for Windows consoles

skeleten opened the pull request:
Unfortunately, the `term` API makes this PR rather inelegant, any hints/pointers to make it better are highly appreciated!

Fixes #2803

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

retep998 commented:
Note that this PR removes support for terminfo consoles on Windows. Ideally what this should do is first check for Windows console support, and then if that fails try terminfo, and only if that also fails fallback to no color.

Also note that regardless of how this code is written, it is still fundamentally flawed due to https://github.com/Stebalien/term/issues/62.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -148,26 +148,55 @@ impl MultiShell {
+                },
+                // This should never happens, since we checked that the creation
+                // would succeed above.
+                Err(e) => unreachable!("Could not create WinConsole: {:?}", e),
```
Could this punt the error upwards as even though the first one succeeded the second one may still have a small window where it could fail?

retep998 replied:
That small window being someone calls `FreeConsole` between the first and second call to `WinConsole::new`.

skeleten replied:
It *is* easy enough though - and If I replace the `unreachable!` with `Shell::get_terminfo_term` the  previous check should be obsolete, it seems more elegant that way to me, actually.

alexcrichton commented:
Thanks @skeleten!

skeleten commented:
I added a commit - Letting the errors bubble up and added a `create_fallback` method that directly creates a `NoColor` term around the given `Box`. This is used by the method `shell` in `cargo/lib.rs` in case the usual creation fails.

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -147,27 +147,61 @@ impl MultiShell {
 
 impl Shell {
-    pub fn create(out: Box<Write + Send>, config: ShellConfig) -> Shell {
+    pub fn create(out: Box<Write + Send>, config: ShellConfig) -> CargoResult<Shell> {
```
This is only ever really called in this one crate, so perhaps the logic of `create_fallback` could be folded into `create` itself?

That is, could we remove `create_fallback`?

skeleten replied:
The problem there is, that the `out` is consumed - It'd needed to be newly created by the caller. As such I'm not really sure if that's at all possible

alexcrichton replied:
Perhaps this could take `FnMut() -> Box<Write + Send>` and just call that multiple times if necessary? The actual handles are just stdout/stderr so they can be manufactured multiple tiems.

skeleten replied:
That sounds like a great idea! I'll be rewriting it in such a manner straight away :)

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -147,27 +147,66 @@ impl MultiShell {
+        Shell::get_terminfo_term(out)
+    }
+
+    fn get_terminfo_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
```
This only returns `Ok`, so it should be fine to remove `CargoResult` from the return type here

alexcrichton commented on src/cargo/core/shell.rs:
```diff
@@ -147,27 +147,66 @@ impl MultiShell {
+
+    /// Create a shell in such a manner that it cannot possibly fail
+    /// Use this after `create` has failed.
+    pub fn create_fallback(out: Box<Write + Send>, config: ShellConfig) -> Shell {
```
This can probably be removed as it's not called.

skeleten replied:
Yea it seems I forgot to append that change to my commit. I'll add these changes!

alexcrichton commented:
Looks good to me, thanks! Can you squash the commits as well?

alexcrichton commented:
@bors: r+ 

Thanks @skeleten!

bors commented:
:pushpin: Commit 6edf535 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 6edf535d975422f832dc9579188fe7df9b81ab03 -->

retep998 commented:
@skeleten You appear to have a merge commit in your PR.

bors commented:
:hourglass: Testing commit 6edf535 with merge 5716f32...

bors commented:
:sunny: Test successful - [cargo-cross-linux](https://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/136), [cargo-linux-32](http://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1692), [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1681), [cargo-mac-32](http://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1688), [cargo-mac-64](http://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1682), [cargo-win-gnu-32](http://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/571), [cargo-win-gnu-64](http://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/571), [cargo-win-msvc-32](http://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/495), [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/522)
